src/builtins.ts

```diff
@@ -2056,6 +2072,279 @@ function builtin_store(ctx: BuiltinContext): ExpressionRef {
 }
 builtins.set(BuiltinNames.store, builtin_store);
 
+// add<T?>(left: T, right: T) -> T
+function builtin_add(ctx: BuiltinContext): ExpressionRef {
+  var compiler = ctx.compiler;
+  var module = compiler.module;
+  if (checkTypeOptional(ctx, true) | checkArgsRequired(ctx, 2))
+    return module.unreachable();
+  var operands = ctx.operands;
+  var typeArguments = ctx.typeArguments;
+  var left = operands[0];
+  var arg0 = typeArguments
+    ? compiler.compileExpression(
+        left,
+        typeArguments[0],
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      )
+    : compiler.compileExpression(operands[0], Type.auto, Constraints.MUST_WRAP);
+  var type = compiler.currentType;
+  if (type.isValue) {
+    let arg1: ExpressionRef;
+    if (!typeArguments && left.isNumericLiteral) {
+      // prefer right type
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.MUST_WRAP
+      );
+      if (compiler.currentType != type) {
+        arg0 = compiler.compileExpression(
+          left,
+          (type = compiler.currentType),
+          Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+        );
+      }
+    } else {
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      );
+    }
+    let op: BinaryOp = -1;
+    switch (type.kind) {
+      case TypeKind.I8:
+      case TypeKind.I16:
+      case TypeKind.U8:
+      case TypeKind.U16:
+      case TypeKind.BOOL: {
+        return compiler.ensureSmallIntegerWrap(
+          module.binary(BinaryOp.AddI32, arg0, arg1),
+          type
+        );
```

Comment:
If I'm not mistaken there is no need to wrap small integers on add, sub and mul. Performing the operation itself does not need wrapped input values (already not enforced here), because the result yields the same bit pattern as for wrapped values except for the high garbage bits that wrapping would truncate. It is therefore fine to leave the result in an unwrapped state (with garbage bits), expecting the next operation that needs a wrapped input value, for example div, to do the wrapping on demand.

Unfortunately there is no proper documentation about this behavior, but as a reference there is `compileBinaryExpression` in compiler.ts, for example doing the following on addition:

https://github.com/AssemblyScript/assemblyscript/blob/c0eb3bb0c6733be13cf76d1946d5134dee418b07/src/compiler.ts#L4579-L4620

So I think that add, sub and mul can be even simpler here, just performing the operation even if types are small integers.

Reply:
Hmm, I'm not sure. Is it would valid for example in this situation?
```ts
assert(u8.add(0xFF, 0xFF) == 0xFE)
```

Reply:
In your example, the `eq` operation would do the wrapping of its input values because it does an exact comparison so needs to truncate garbage bits for the comparison to be correct. In previous conversations I titled the behavior "lazy wrapping" I believe, and it was introduced in the early days when the problem of too much unnecessary (eager) wrapping was brought up.

Reply:
Hmm, so may be `ensureSmallIntegerWrap` unnecessary in rest places as well? Or it's only valid for builtins?

Reply:
Yes, I think `ensureSmallIntegerWrap` is not necessary on the `add`, `sub` and `mul` builtins, and can align to what the compiler does on `compileBinaryExpression`. Note that an `eq` builtin, if it is added, needs to compile the LHS and RHS with `Constraints.MUST_WRAP`, or set `wrap=true` in `convertExpression`, or needs to use `ensureSmallIntegerWrap`. That's all kinda the same and just differs in when the wrapping is performed.

Reply:
Makes me wonder if we can refactor that so the compiler just provides `compileAdd(left, leftType, right, rightType)` instead of inlining the switch, so `compileAdd` can be used in the builtin. Hmm.

Reply:
That's will be much cleaner.

Reply:
Ok, in this case I think it is fine to merge this PR and you or I take a look at refactoring in a follow-up? Does that sound fine?

Reply:
I'm agree (If you ask it to me)

Reply:
Is there anything I need to do in this PR?

Reply:
No, I think it is fine. It also helped us to realize that there is some refactoring we can do to make the process easier in the future :)